61-matamata-turtle/technical_specification.md

```diff
@@ -0,0 +1,247 @@
+
+The event ID is always created, but only propagated if it enters the DLQ cycle.
+This differs from the correlation ID, which is propagated across services and persisted
+in the database in the case of outbox events.
+
+To illustrate:
+
+![Event ID usage](./images/event%20ID.png)
+1. An event is published to the "users" topic, where it is stored at partition 0, offset 17.
+2. The `hexkit` Kafka provider used by the `NOS` gets the message from `aiokafka`.
+3. The Kafka provider hands the transformed event over to the service-defined
+   translator, which performs some business logic driven by the event.
+4. The translator encounters an error which it cannot mitigate, and all retries fail.
+5. The Kafka provider logs the error, including the event ID (`"nos,users,0,17"`).
+6. The Kafka provider publishes the event with the event ID header to the DLQ topic.
+   - The original topic header is also included.
+7. We use the DLQ service to take a look at the failed event.
+8. We learn the problem was actually a database issue that we've since fixed.
+9.  We use the DLQ service to publish the event to the retry topic, sans event ID header.
+   Instead, the DLQ service includes a special header with the original topic ("users").
+10. The `NOS` encounters the republished event, this time from its dedicated retry
+  topic. The Kafka provider understands that the retry topic is special, so it obtains
+  the topic field from the original topic header. If the retry event fails again, the
+  old event ID would be misleading had we modified the event in the DLQ service.
+
+If the translator encounters no errors, the event ID is not needed and never created.
+
+
+### Event processing:
+When we resolve the next event in a given DLQ topic, it can go one of two ways:
+- We can discard the event, meaning it is not republished to a retry queue and is
+  effectively ignored.
+- We can process the event, where the event is published to the retry topic or
+  discarded (e.g. if the service info is missing). The processing can also be bypassed
+  by directly supplying a JSON event to the API, which publishes the event outright
+  as long as the correlation ID matches.
+
+Another useful piece of information for the DLQ service is the exception information.
+There is currently no mechanism to provide that to the DLQ service, but the info could
+be passed through event headers. The exception header would then be removed upon
+republishing the event to a retry topic.
+
+This [Java DLQ implementation](https://medium.com/nerd-for-tech/-to-re-queue-apache-kafka-dlq-messages-95941525ca77)
+uses such headers (toward the bottom).
+
+### Processing Dry Run:
+It would be nice to be able to perform a dry-run of event processing as a sanity check.
+The test endpoint would perform any applicable processing on the next event and
+return the result instead of publishing it. If processing would result in the
+event being discarded instead of published, then it will return nothing.  
```

Comment:
Above you already mention this as a feature to be implemented instead of a nice to have. Maybe just remove this section.